Introduce a shared scm context

Mat-Co opened the pull request:
In some situations, when many local cookbooks are used, tests can spend a lot of time doing several times the same git commands for each one of them. It can be particularly inefficient when they share the same git repo.

With this change, we make sure commands are launched only once and results are reused if necessary.

## Description
For each local cookbook, browse the directory tree and look for a git configuration. This scm context is then re-used for other local cookbooks that are sharing the same git configuration folder.

Commit:
Introduce a shared scm context

In some situations, when many local cookbooks are used, tests can
spend a lot of time doing several times the same git commands for each
one of them. It can be particularly unefficient when they share the same
git repo.

With this change, we make sure commands are launched only once and
results are reused if necessary.

Commit:
Revert unnecessary change on CookbookLock

SCM context is used for local cookbooks only

Annih commented on lib/chef-cli/cookbook_profiler/git.rb:
```diff
@@ -103,6 +104,17 @@ def current_branch
       private
+      
+      def repo_directory
+        @repo_directory ||=
```
Indentation is weird, the following more compact form seems clearer to me and fix the various other comments

```ruby
    def repo_directory
      @repo_directory ||= Pathname.new(@cookbook_path).ascend do |parent_dir|
        possible_git_dir = File.join(parent_dir, '.git')
        break possible_git_dir if File.directory?(possible_git_dir )
      end
    end
```

Mat-Co replied:
Done, thx for the suggestion !

Annih commented on lib/chef-cli/cookbook_profiler/git.rb:
```diff
@@ -103,6 +104,17 @@ def current_branch
+              return possbile_git_dir
+            end
+          end
+          @repo_directory
```
Useless since `@repo_directory ||= XXX` returns the value

Mat-Co replied:
Done

Annih commented on lib/chef-cli/cookbook_profiler/git.rb:
```diff
@@ -103,6 +104,17 @@ def current_branch
+          Pathname.new(@cookbook_path).ascend do |parent_dir|
+            possbile_git_dir = File.join(parent_dir, '.git')
+            if File.directory?(possbile_git_dir)
+              return possbile_git_dir
```
Don't use return from a block, it bypass everything (especially the `@xxx ||=` )

Mat-Co replied:
Done

Annih commented on lib/chef-cli/cookbook_profiler/git.rb:
```diff
@@ -103,6 +104,17 @@ def current_branch
+      def repo_directory
+        @repo_directory ||=
+          Pathname.new(@cookbook_path).ascend do |parent_dir|
+            possbile_git_dir = File.join(parent_dir, '.git')
```
Typo: `possbile_git_dir` -> `possible_git_dir`

Mat-Co replied:
Done

Commit:
Refactor repo_directory method

As discussed in the scm context pull request, there were different
small issues with the previous implem.
